File: server/src/main/java/com/cts/mfrp/procuresphere/service/UserService.java
package com.cts.mfrp.procuresphere.service;

import com.cts.mfrp.procuresphere.dto.request.ProfileUpdateRequest;
import com.cts.mfrp.procuresphere.dto.response.UserResponse;
import com.cts.mfrp.procuresphere.exception.BadRequestException;
import com.cts.mfrp.procuresphere.exception.ResourceNotFoundException;
import com.cts.mfrp.procuresphere.model.Role;
import com.cts.mfrp.procuresphere.model.User;
import com.cts.mfrp.procuresphere.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
public class UserService {

    private final UserRepository userRepository;

    public List<UserResponse> getAll() {
        return userRepository.findAll().stream().map(this::toResponse).toList();
    }

    public UserResponse getById(Long id) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("User not found with id: " + id));
        return toResponse(user);
    }

    public UserResponse getByEmail(String email) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new ResourceNotFoundException("User not found with email: " + email));
        return toResponse(user);
    }

    @Transactional
    public UserResponse updateRole(Long id, String roleStr) {
        if (roleStr == null || roleStr.isBlank()) {
            throw new BadRequestException("Role is required");
        }
        Role role;
        try {
            role = Role.valueOf(roleStr.toUpperCase());
        } catch (IllegalArgumentException e) {
            throw new BadRequestException("Invalid role: " + roleStr + ". Valid values: ADMIN, MANAGER, USER, SUPPLIER");
        }
        User user = userRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("User not found with id: " + id));
        user.setRole(role);
        userRepository.save(user);
        return toResponse(user);
    }

    /** Self-service profile update. Used by PUT /api/users/me. */
    @Transactional
    public UserResponse updateProfile(String currentEmail, ProfileUpdateRequest req) {
        User user = userRepository.findByEmail(currentEmail)
                .orElseThrow(() -> new ResourceNotFoundException("User not found: " + currentEmail));

        // If email is changing, ensure uniqueness against OTHER users.
        if (req.getEmail() != null && !req.getEmail().equalsIgnoreCase(user.getEmail())) {
            if (userRepository.existsByEmail(req.getEmail())) {
                throw new BadRequestException("Email '" + req.getEmail() + "' is already in use");
            }
            user.setEmail(req.getEmail());
        }

        if (req.getFirstName() != null)  user.setFirstName(req.getFirstName());
        if (req.getLastName()  != null)  user.setLastName(req.getLastName());
        if (req.getPhone()     != null)  user.setPhone(req.getPhone());
        if (req.getDepartment()!= null)  user.setDepartment(req.getDepartment());
        if (req.getCompany()   != null)  user.setCompany(req.getCompany());
        if (req.getAddress()   != null)  user.setAddress(req.getAddress());
        if (req.getAvatarUrl() != null)  user.setAvatarUrl(req.getAvatarUrl());

        return toResponse(userRepository.save(user));
    }

    @Transactional
    public void delete(Long id) {
        if (!userRepository.existsById(id)) {
            throw new ResourceNotFoundException("User not found with id: " + id);
        }
        userRepository.deleteById(id);
    }

    private UserResponse toResponse(User user) {
        return UserResponse.builder()
                .userId(user.getUserId())
                .firstName(user.getFirstName())
                .lastName(user.getLastName())
                .email(user.getEmail())
                .phone(user.getPhone())
                .role(user.getRole())
                .department(user.getDepartment())
                .company(user.getCompany())
                .address(user.getAddress())
                .avatarUrl(user.getAvatarUrl())
                .build();
    }
}

File: Procosphere_Angular/src/app/services/user.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';

export interface UserResponse {
  userId: number;
  firstName: string;
  lastName: string;
  email: string;
  phone: string | null;
  role: 'ADMIN' | 'MANAGER' | 'USER' | 'SUPPLIER';
  department: string | null;
  company: string | null;
  address: string | null;
  avatarUrl: string | null;
}

export interface ProfileUpdateRequest {
  firstName?: string;
  lastName?: string;
  email?: string;
  phone?: string;
  department?: string;
  company?: string;
  address?: string;
  avatarUrl?: string;
}

export interface ApiResponse<T> { success: boolean; message: string; data: T; }

@Injectable({ providedIn: 'root' })
export class UserService {
  private base = environment.apiUrl + '/api/users';
  constructor(private http: HttpClient) {}

  getMe(): Observable<ApiResponse<UserResponse>> {
    return this.http.get<ApiResponse<UserResponse>>(`${this.base}/me`);
  }

  updateMe(req: ProfileUpdateRequest): Observable<ApiResponse<UserResponse>> {
    return this.http.put<ApiResponse<UserResponse>>(`${this.base}/me`, req);
  }

  getAll(): Observable<ApiResponse<UserResponse[]>> {
    return this.http.get<ApiResponse<UserResponse[]>>(this.base);
  }

  updateRole(id: number, role: string): Observable<ApiResponse<UserResponse>> {
    return this.http.patch<ApiResponse<UserResponse>>(`${this.base}/${id}/role`, { role });
  }

  delete(id: number): Observable<ApiResponse<void>> {
    return this.http.delete<ApiResponse<void>>(`${this.base}/${id}`);
  }
}
